test/integration/boundingbox_camera_plugin.cc

```diff
@@ -0,0 +1,474 @@
+    g_boxes.push_back(box);
+  }
+  g_counter++;
+  g_mutex.unlock();
+}
+
+/// \brief callback to receive 3D boxes from the camera
+void OnNew3DBoundingBoxes(const msgs::AnnotatedOriented3DBox_V &boxes)
+{
+  g_mutex.lock();
+  g_boxes.clear();
+
+  int size = boxes.annotated_box_size();
+  for (int i = 0; i < size; i++)
+  {
+    BoundingBox box(BoundingBoxType::Box3D);
+    auto annotated_box = boxes.annotated_box(i);
+    auto orientedBox = annotated_box.box();
+    // center
+    box.center.Set(orientedBox.center().x(), orientedBox.center().y(),
+      orientedBox.center().z());
+    // orientation
+    auto orientation = orientedBox.orientation();
+    box.oreintation.Set(orientation.w(), orientation.x(),
+      orientation.y(), orientation.z());
+    // size
+    auto boxSize = orientedBox.boxsize();
+    box.size.Set(boxSize.x(), boxSize.y(), boxSize.z());
+
+    box.label = annotated_box.label();
+    g_boxes.push_back(box);
+  }
+  g_counter++;
+  g_mutex.unlock();
+}
+
+/// \brief wait till you read the published frame
+void WaitForNewFrame()
+{
+  auto waitTime = std::chrono::duration_cast< std::chrono::milliseconds >(
+  std::chrono::duration< double >(0.001));
+  int counter = 0;
+  // wait for the counter to increase
+  for (int sleep = 0; sleep < 300 && counter == 0; ++sleep)
+  {
+    g_mutex.lock();
+    counter = g_counter;
+    g_mutex.unlock();
+    std::this_thread::sleep_for(waitTime);
+  }
+  EXPECT_GT(counter, 0);
+}
+
+/// \brief Build a scene with 3 boxes 2 overlapping boxes which 1
+/// is behind the other, and the 3rd box is invisible behind them
+void BuildScene(rendering::ScenePtr scene)
+{
+  math::Vector3d occludedPosition(4, 1, 0);
+  math::Vector3d frontPosition(2, 0, 0);
+  math::Vector3d invisiblePosition(5, 0, 0);
+
+  rendering::VisualPtr root = scene->RootVisual();
+
+  // create front box visual (the smaller box)
+  rendering::VisualPtr occludedBox = scene->CreateVisual();
+  occludedBox->AddGeometry(scene->CreateBox());
+  occludedBox->SetOrigin(0.0, 0.0, 0.0);
+  occludedBox->SetLocalPosition(occludedPosition);
+  occludedBox->SetLocalRotation(0, 0, 0);
+  occludedBox->SetUserData("label", 1);
+  root->AddChild(occludedBox);
+
+  // create occluded box visual (the bigger box)
+  rendering::VisualPtr frontBox = scene->CreateVisual();
+  frontBox->AddGeometry(scene->CreateBox());
+  frontBox->SetOrigin(0.0, 0.0, 0.0);
+  frontBox->SetLocalPosition(frontPosition);
+  frontBox->SetLocalRotation(0, 0, 0);
+  frontBox->SetUserData("label", 2);
+  root->AddChild(frontBox);
+
+  rendering::VisualPtr invisibleBox = scene->CreateVisual();
+  invisibleBox->AddGeometry(scene->CreateBox());
+  invisibleBox->SetOrigin(0.0, 0.0, 0.0);
+  invisibleBox->SetLocalPosition(invisiblePosition);
+  invisibleBox->SetLocalRotation(0, 0, 0);
+  invisibleBox->SetUserData("label", 2);
+  root->AddChild(invisibleBox);
+}
+
+/// \brief Build a scene with 3d oreinted box
+void BuildScene3D(rendering::ScenePtr scene)
```

Comment:
Would it be possible to add a model with multiple links to the 3d oriented box test so that we can make sure the bounding boxes for each link get combined/merged correctly?

Reply:
how can we do that .. through rendering APIs ??